### DISCUSSION TITLE
Supporting Tiled and/or LDtk

### DISCUSSION THREAD

**Discussion by davesmith00000** [PurpleKingdomGames/indigo#270] [Collaborations]
There are a few separate people who have either started trying to improve support for Tiled in Indigo, or who have expressed an interest in doing so. I'd like to try and pull them all together to see if we can get this done together, as it's a topic that keeps coming up.

There are a number of tile map editors out there of interest. The two of interest are:

1. Tiled - some support already - https://www.mapeditor.org/
2. LDtk - not supported at all yet - https://ldtk.io/

Are there any others?

**Comment by davesmith00000** [comment]
People known or suspected of being interested. 😄 
@laurensdv
@sylvansson
@sintrastes
@agdt3
@hobnob

**Comment by davesmith00000** [comment]
Relevant issues:
https://github.com/PurpleKingdomGames/indigo/issues/67
https://github.com/PurpleKingdomGames/indigo/issues/246

**Comment by davesmith00000** [comment]
I think the first useful thing to do would be to take stock of where we are now.

Indigo comes with a very basic Tiled importer that I wrote - just enough - to make "The Cursed Pirate" demo. You can see it working in this example:
https://github.com/PurpleKingdomGames/indigo-examples/blob/master/examples/tiled/src/main/scala/indigoexamples/TiledExample.scala

Separately @laurensdv has been working on a new version, here:
https://github.com/laurensdv/indigo/blob/a5497d880a8fc7e73f4df691d42eb9a54cff6181/indigo/indigo/src/main/scala/indigo/shared/formats/TiledMap.scala

Could you tell us how you're getting on?

**Reply by laurensdv** [reply]
@davesmith00000 In the meantime I am focussing to keep my Indigo game code up to date with new releases as well as keeping the TiledMap with animation fork/branch in sync. However still having the GL errors with the clone approach with animated tiles (sprites) as #246. Regular approach (without clones) still works but very CPU intensive as mentioned as well in #246.

Shifting my attention to continue developing the demo game content as well in the past few weeks and ignoring the issue for now until resolution becomes clear.

@davesmith00000 It'd be good to see the TiledMap with animations branch integrated here - would be easier for future development, but then I understand unit tests would need to be implemented as next step. I would be happy to contribute on that part. What are now the plans for (animated) tilemap (with clones) support?

**Reply by davesmith00000** [reply]
I'm trying to work out what the plan is.

I'm going to get the next release out soon, because there's lots of bits and pieces that have landed and Indigo has become a dependancy for Tyrian.

However after that I think I'd like to try and reconcile all the various efforts / wishes to improve Tiled if I/we can, I just don't want to make it a blocker for the current work. It helps that I think I'm going to need the Tiled support myself, so there's a bit of extra motivation there. 😁

**Reply by davesmith00000** [reply]
And of course, I'd really like you and the work you've been doing to be part of that!

**Comment by zetashift** [comment]
I'm very much interested in writing something for LDtk in Scala, past days I've been looking at other importers for LDtk and I might need some guidance on how this works because of the following reasons: I don't know anything about game tooling and I don't know much about Scala & JSON, except for some uPickle stuff. 

I want to write a library following this route: https://ldtk.io/docs/game-dev/loading-ldtk-in-your-game/#3-the-home-made-way
I'll try to get a repo up from the progress I've made so far

**Reply by davesmith00000** [reply]
Hi @zetashift - great to have you aboard!

Oh yeah - the docs look good. Currently Indigo is using Circe but we used to support uPickle too. Circe _is_ slightly easier when dealing with more complicated types I must say, from experience. Might be worth experimenting with it?

Essentially, all we need for any implementation is:

1. A text file of the level data
2. The asset files of images / textures that go with it
3. Code that can parse the data into a nice Scala description
4. Code the converts the parsed version into something that Indigo understands, which is actually two things: A model representation (non-graphic, something that allows you to decide if your character is standing on a platform or not) and a renderable version. But I would focus on the renderable version initially as it's the most fun.

**Reply by zetashift** [reply]
I'm gonna experiment with Circe! Let's see how far I can get haha.

**Comment by davesmith00000** [comment]
It might be worth breaking these out into separate libraries rather than them living in the main Indigo repo. Anyone have an opinion?

**Reply by Sintrastes** [reply]
In all of the multi-artifact projects I've worked on, I always find monorepos the most convenient to work with rather than separate repos, but in terms of splitting these into different artifacts/libraries -- that makes sense to me.
